userspace/otpilot/src/gpio.rs

```diff
@@ -0,0 +1,217 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use core::cell::Cell;
+use core::convert::TryFrom;
+
+use libtock::gpio::GpioPinUnitialized;
+use libtock::gpio::GpioPinRead;
+use libtock::gpio::GpioPinWrite;
+use libtock::gpio::{IrqMode, InputMode};
+use libtock::result::TockError;
+use libtock::result::TockResult;
+
+/// GPIO pins.
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+#[allow(non_camel_case_types)]
+pub enum GpioPin {
+    BMC_SRST_N = 0,
+    BMC_CPU_RST_N = 1,
+    SYS_RSTMON_N = 2,
+    BMC_RSTMON_N = 3,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+#[allow(dead_code)]
+pub enum GpioValue {
+    Low,
+    High,
+}
+
+pub trait GpioControl {
+    /// Check if there are any events to be consumed.
+    fn have_events(&self) -> bool;
+
+    /// Consume one event on the specified pin.
+    /// Returns true if there was an event to be consumed.
+    fn consume_event(&self, pin: GpioPin) -> bool;
+
+    /// Clear all events on the specified pin.
+    /// Returns true if there was an event.
+    fn clear_event(&self, pin: GpioPin) -> bool;
+
+    /// Set GpioPin value.
+    fn set(&self, pin: GpioPin, val: GpioValue) -> TockResult<()>;
+}
+
+// Get the static GpioControl object.
+pub fn get() -> &'static dyn GpioControl {
+    get_impl()
+}
+
+/// Error for invalid GpioPin conversion.
+pub struct InvalidGpioPin;
+
+impl TryFrom<usize> for GpioPin {
+    type Error = InvalidGpioPin;
+
+    fn try_from(item: usize) -> Result<GpioPin, Self::Error> {
+        match item {
+            0 => Ok(GpioPin::BMC_SRST_N),
+            1 => Ok(GpioPin::BMC_CPU_RST_N),
+            2 => Ok(GpioPin::SYS_RSTMON_N),
+            3 => Ok(GpioPin::BMC_RSTMON_N),
+            _ => Err(InvalidGpioPin),
+        }
+    }
+}
+
+impl From<GpioPin> for usize {
+    fn from(item: GpioPin) -> usize {
+        item as usize
+    }
+}
+
+
+struct GpioControlImpl {
+    bmc_srst_n: Option<GpioPinWrite>,
+    bmc_cpu_rst_n: Option<GpioPinWrite>,
+    sys_rstmon_n: Option<GpioPinRead>,
+    bmc_rstmon_n: Option<GpioPinRead>,
+
+    sys_rstmon_n_events: Cell<usize>,
+    bmc_rstmon_n_events: Cell<usize>,
+}
+
+static mut GPIO_CTRL: GpioControlImpl = GpioControlImpl {
+    bmc_srst_n: None,
+    bmc_cpu_rst_n: None,
+    sys_rstmon_n: None,
+    bmc_rstmon_n: None,
+    sys_rstmon_n_events: Cell::new(0),
+    bmc_rstmon_n_events: Cell::new(0),
+};
+
+static mut IS_INITIALIZED: bool = false;
+
+fn get_impl() -> &'static GpioControlImpl {
+    unsafe {
+        if !IS_INITIALIZED {
+            if GPIO_CTRL.initialize().is_err() {
+                panic!("Could not initialize GPIO Control");
+            }
+            IS_INITIALIZED = true;
+        }
+        &GPIO_CTRL
+    }
+}
+
+
+fn add_event(events: &Cell<usize>) {
+    let val = events.get();
+    events.set(val + 1);
+}
+
+fn delete_event(events: &Cell<usize>) -> usize {
+    let val = events.get();
+    if val > 0 {
```

Comment:
Hmm.  Should it be OK to silently do nothing here when `val == 0`?

Reply:
I added a comment explaining what this method is doing. So yes, that's ok.